compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1676,20 +1747,36 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         debug!(?captures);
 
         self.with_hir_id_owner(opaque_ty_node_id, |this| {
-            let future_bound =
-                this.while_capturing_lifetimes(opaque_ty_def_id, &mut captures, |this| {
-                    // We have to be careful to get elision right here. The
-                    // idea is that we create a lifetime parameter for each
-                    // lifetime in the return type.  So, given a return type
-                    // like `async fn foo(..) -> &[&u32]`, we lower to `impl
-                    // Future<Output = &'1 [ &'2 u32 ]>`.
-                    //
-                    // Then, we will create `fn foo(..) -> Foo<'_, '_>`, and
-                    // hence the elision takes place at the fn site.
-                    this.lower_async_fn_output_type_to_future_bound(output, fn_def_id, span)
-                });
-            debug!("lower_async_fn_ret_ty: future_bound={:#?}", future_bound);
-            debug!("lower_async_fn_ret_ty: captures={:#?}", captures);
+            let lifetime_stash = std::mem::replace(
+                &mut this.captured_lifetimes,
+                Some(LifetimeCaptureContext {
+                    parent_def_id: opaque_ty_def_id,
+                    captures: std::mem::take(&mut captures),
+                    binders_to_ignore: Default::default(),
+                }),
+            );
+
+            let lifetimes_in_bounds =
+                lifetime_collector::lifetimes_in_ret_ty(&this.resolver, output);
+            debug!(?lifetimes_in_bounds);
+
+            this.create_and_capture_lifetime_defs(&lifetimes_in_bounds);
+
+            // We have to be careful to get elision right here. The
+            // idea is that we create a lifetime parameter for each
+            // lifetime in the return type.  So, given a return type
+            // like `async fn foo(..) -> &[&u32]`, we lower to `impl
+            // Future<Output = &'1 [ &'2 u32 ]>`.
+            //
+            // Then, we will create `fn foo(..) -> Foo<'_, '_>`, and
+            // hence the elision takes place at the fn site.
```

Comment:
This comment is likely out of date, right? Elision is happening in an earlier phase now, I think.

cc @cjgillot